File: app/src/main/java/com/itachi1706/cheesecakeutilities/Modules/VehicleMileageTracker/FirebaseUtils.java
package com.itachi1706.cheesecakeutilities.Modules.VehicleMileageTracker;

import com.google.firebase.database.FirebaseDatabase;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Created by Kenneth on 28/7/2017.
 * for com.itachi1706.cheesecakeutilities.Modules.VehicleMileageTracker in CheesecakeUtilities
 */

public class FirebaseUtils {

    // Constants
    public static final String FB_REC_USER = "users", FB_REC_STATS = "statistics", FB_REC_RECORDS = "records";

    private static FirebaseDatabase firebaseDatabase;

    static final int RECORDS_VERSION = 3;

    public static FirebaseDatabase getFirebaseDatabase() {
        if (firebaseDatabase == null) {
            firebaseDatabase = FirebaseDatabase.getInstance();
            firebaseDatabase.setPersistenceEnabled(true);
        }
        return firebaseDatabase;
    }

    public static String formatTime(long time) {
        return formatTime(time, "dd MMMM yyyy HH:mm");
    }

    public static String formatTime(long time, String format) {
        SimpleDateFormat sdf = new SimpleDateFormat(format, Locale.US);
        Date dt = new Date();
        dt.setTime(time);
        return sdf.format(dt);
    }

    public static String formatTimeDuration(long start, long end) {
        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yy HHmm", Locale.US);
        Date dt = new Date();
        dt.setTime(start);
        String timeString = sdf.format(dt);
        sdf.applyPattern("dd/MM/yy HHmm zzz");
        dt.setTime(end);
        timeString += " - " + sdf.format(dt);
        return timeString;
    }
}

File: app/src/main/java/com/itachi1706/cheesecakeutilities/util/FirebaseUtils.kt
package com.itachi1706.cheesecakeutilities.util

import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.math.roundToLong

/**
 * Created by Kenneth on 3/6/2019.
 * for com.itachi1706.cheesecakeutilities.Util in CheesecakeUtilities
 */
abstract class FirebaseUtils {

    companion object {

        private var firebaseDatabase: FirebaseDatabase? = null

        fun removeListener(listener: ValueEventListener) {
            getFirebaseDatabase().reference.removeEventListener(listener)
        }

        fun getFirebaseDatabase(): FirebaseDatabase {
            if (firebaseDatabase == null) {
                firebaseDatabase = FirebaseDatabase.getInstance()
                firebaseDatabase!!.setPersistenceEnabled(true)
            }
            return firebaseDatabase!!
        }

        fun getDatabaseReference(tag: String): DatabaseReference {
            val database = getFirebaseDatabase()
            return getDatabaseReference(tag, database)
        }

        fun getDatabaseReference(tag: String, database: FirebaseDatabase): DatabaseReference {
            return database.reference.child(tag)
        }

        fun formatTime(time: Long): String {
            return formatTime(time, "dd MMMM yyyy HH:mm")
        }

        fun formatTime(time: Long, format: String): String {
            val sdf = SimpleDateFormat(format, Locale.US)
            val dt = Date()
            dt.time = time
            return sdf.format(dt)
        }

        fun formatTimeDuration(start: Long, end: Long): String {
            val sdf = SimpleDateFormat("dd/MM/yy HHmm", Locale.US)
            val dt = Date()
            dt.time = start
            var timeString = sdf.format(dt)
            sdf.applyPattern("dd/MM/yy HHmm zzz")
            dt.time = end
            timeString += " - " + sdf.format(dt)
            return timeString
        }

        fun parseData(d: Double, decimal: Boolean): String {
            return if (decimal) String.format(Locale.getDefault(), "%.1f", d) else
                String.format(Locale.getDefault(), "%d", d.roundToLong())
        }
    }
}
